crates/codegen/schema/src/validation/rules/empty_productions.rs

```diff
@@ -26,86 +25,140 @@ impl EmptyProductionsVisitor {
 }
 
 impl Visitor for EmptyProductionsVisitor {
-    fn visit_production(
-        &mut self,
-        production: &ProductionRef,
-        _reporter: &mut Reporter,
-    ) -> VisitorResponse {
-        match production.kind.value {
-            ProductionKind::Trivia => {
-                // Skip. Allowed to be empty.
-                return VisitorResponse::StepOut;
-            }
-            ProductionKind::Token | ProductionKind::Rule => {
-                // Check the root expression.
-                return VisitorResponse::StepIn;
-            }
-        };
+    fn visit_scanner(&mut self, scanner: &ScannerRef, reporter: &mut Reporter) -> VisitorResponse {
+        if let Some(cst_node) = find_empty_scanner_node(scanner) {
+            reporter.report(cst_node, Errors::PossibleEmptyScanner);
+        }
+
+        // Only check the top-most expression. Ignore nested ones.
+        return VisitorResponse::StepOut;
     }
 
-    fn visit_expression(
-        &mut self,
-        expression: &ExpressionRef,
-        reporter: &mut Reporter,
-    ) -> VisitorResponse {
-        if let Some(syntax) = find_empty_node(expression) {
-            reporter.report(syntax, Errors::PossibleEmptyRoot);
+    fn visit_parser(&mut self, parser: &ParserRef, reporter: &mut Reporter) -> VisitorResponse {
+        if let Some(cst_node) = find_empty_parser_node(parser) {
+            reporter.report(cst_node, Errors::PossibleEmptyParser);
         }
 
         // Only check the top-most expression. Ignore nested ones.
         return VisitorResponse::StepOut;
     }
+
+    fn visit_precedence_parser(
+        &mut self,
+        _: &PrecedenceParserRef,
+        _: &mut Reporter,
+    ) -> VisitorResponse {
+        // To ensure we don't visit the operator definitions, we return StepOut.
```

Comment:
Not sure why are we skipping operators? Shouldn't we be making sure that they are not empty? For example, a potentially empty `UnarySuffix` that won't consume characters.

Reply:
Similarly for the primary expressions.